### DISCUSSION TITLE
Decision procedures for decidable fragments of quantified FO

### DISCUSSION THREAD

**Discussion by nicola-gigante** [cvc5/cvc5#8770] [Q&A] [answered]
What are the fragments of quantified formulas (with uninterpreted symbols) for which CVC5 guarantees termination?

For example, does it recognise and act accordingly for formulas included in the Bernays-Schönfinkel class (∃&#42;∀&#42;), or the Ackermann class (∃&#42;∀&#42;∃&#42;), or similar decidable fragments? 

For example, as far as I know, Z3 is guaranteed to terminate on EPR formulas (aka Bernays-Schönfinkel).

**Comment by ajreynol** [comment, marked as answer]
Hi,
Sorry for the delay in reply, I misplaced this discussion.

There are several classes of quantified formulas that cvc5 is terminating for:

(1) Bernays-Schönfinkel, when using the option `--finite-model-find` (https://homepage.divms.uiowa.edu/~ajreynol/cade24.pdf).

(2) Quantified linear int/real arithmetic (with arbitrary nested quantification), where cvc5 uses techniques from https://homepage.divms.uiowa.edu/~ajreynol/fmsd17-instla.pdf by default.

(3) Finite interpreted quantification, e.g. quantifiers over bitvectors (which also can be arbitrarily nested).  See e.g. https://homepage.divms.uiowa.edu/~ajreynol/cav18.pdf for the techniques cvc5 uses; these techniques are used by default for quantifiers over e.g. BV when other theories are not present. They can be forced on using the option `--cegqi-all`.

For each of these, cvc5 uses techniques that are based around quantifier instantiation. At a high level, the termination argument for each is that only a finite number of quantifier instantiations will be considered, and it always makes progress towards enumerating them.

Of course, these are only theoretical termination arguments, where in practice the number of instantiations may be infeasible to consider. Nevertheless, since quantifier instantiation is done lazily, cvc5 may terminate quickly (for both "sat" and "unsat") in cases where a model (resp. refutation) is found.

If you do have a benchmark where cvc5 is answering "unknown" that you believe cvc5 should be terminating, let me know. 

Hope this helps.

**Reply by nicola-gigante** [reply]
Thanks for the reply! Regarding number (2), do you mean only pure int/real arithmetic formulas or can arithmetic also be mixed with EUF predicates/functions, eg can we mix LIA with Bernays-Schönfinkel? 

BTW, are you at FLoC right now? It would be great to talk with you in-person about these things.

**Reply by ajreynol** [reply]
For (2), this covers only pure int/real arithmetic formulas.  In principle, the two results could be combined to cover the combination, however, the implementation in cvc5 is not set up to handle this.

Unfortunately I won't be at FLoC this time.  However, other members of the cvc5 team will be, including those who are experts in the above techniques @HanielB @aniemetz @mpreiner .

**Reply by nicola-gigante** [reply]
Thanks! Just another question: the Bernays-Schonfinkel fragment has the exists/forall structure. But is implication in the syntax? I mean, if I have A -> B and both A and B are in the fragment, is the implication in the fragment where cvc5 terminates? If the fragment does not formally include the implication, then the above formula is !A or B, but !A is not in the fragment anymore because it becomes of the form forall/exists. 

So is cvc5 terminating also if I use implications and double implications freely?

I hope my question is clear enough, I’m writing from my phone right now (will edit later to be clearer)

**Comment by ajreynol** [comment]
Internally, `(exists x F)` is syntax sugar for `(not (forall x (not F)))` in cvc5; similarly `(=> F1 F2)` is effectively the same as `(or (not F1) F2)`.

Thus, Bernays-Schonfinkel fragment internally is then any input formulas where `forall` is not nested inside of other `forall`, and where there are no (non-constant UF), where note that free declare-fun are implicitly the outmost extential quantification in this setting. However, apart from beneath other quantifers, `forall` can occur anywhere in the overall input (including under negations).

Hope this helps.

**Reply by nicola-gigante** [reply]
Yes, thank you, your answers help a lot!
